### DISCUSSION TITLE
Call another plugin from inside plugin

### DISCUSSION THREAD

**Discussion by melloware** [webpack/webpack#15071] [Q&A] [answered]
I have a plugin I am modifying.  I would like to be able to add an Environment Variable inside my plugin like this using the `DefinePlugin`...

```javascript
      const definitions = {};
      definitions[`process.env.REACT_APP_CSS_NONCE`] = JSON.stringify(this.nonce);
      new webpack.DefinePlugin(definitions).apply(compiler);
```

I have verified the definitions collection is correct but I can't seem to get it to inject my environment variable.  I was following what the `Environment` plugin was doing but can't quite get it to work.

Is this the correct approach to call a plugin from inside another plugin?

**Comment by alexander-akait** [comment]
Where you put this code?

**Reply by melloware** [reply]
I have tried it in a couple of places including the `apply` method. Is there a correct place this code should go?

**Reply by alexander-akait** [reply]
`apply` is correct place and should work, maybe you can provide example of the problem (github repo)?

**Reply by melloware** [reply]
Thank you!  OK let me debug some more and then I will post my GitHub code.

**Reply by melloware** [reply]
OK here is what I am doing...

```javascript
 /**
   * Hooks into webpack to collect assets and hash them, build the policy, and add it into our HTML template
   * @param compiler
   */
  apply(compiler) {
    compiler.hooks.compilation.tap('CspHtmlWebpackPlugin', (compilation) => {
      HtmlWebpackPlugin.getHooks(compilation).beforeAssetTagGeneration.tapAsync(
        'CspHtmlWebpackPlugin',
        this.mergeOptions.bind(this, compilation)
      );
      HtmlWebpackPlugin.getHooks(compilation).beforeEmit.tapAsync(
        'CspHtmlWebpackPlugin',
        this.processCsp.bind(this, compilation)
      );
      HtmlWebpackPlugin.getHooks(compilation).beforeAssetTagGeneration.tapAsync(
        'CspHtmlWebpackPlugin',
        this.getFilesToHash.bind(this)
      );
      HtmlWebpackPlugin.getHooks(compilation).alterAssetTags.tapAsync(
        'CspHtmlWebpackPlugin',
        this.addIntegrityAttributes.bind(this, compilation)
      );
    });

      const definitions = {};
      definitions[`process.env.REACT_APP_CSS_NONCE`] = JSON.stringify(this.primeReactInlineNonce);
      console.log(definitions);
      new webpack.DefinePlugin(definitions).apply(compiler);
  }
```

but my logging comes out like this...

```
Creating an optimized production build...
{ 'process.env.REACT_APP_CSS_NONCE': undefined }
[
  "'nonce-4aa+wcGcaTqhab3A17Y9VA=='",
  "'nonce-/pww3NJJQjxup3ODYLx+bw=='"
]
Compiled successfully.
```

So the problem is its calling the apply method immediately but my code calculating the value for NONCE is happening asynchronously.  How is it possible for me to gather a value as part of my build and then output it in the DefinePlugin?

**Reply by alexander-akait** [reply]
Unfortunately, it is too late to apply `DefinePlugin`, because `DefinePlugin` https://github.com/webpack/webpack/blob/main/lib/DefinePlugin.js#L280 use the same hook (i.e. `compiler.hooks.compilation`). Anyway we have very closed case in our tests https://github.com/webpack/webpack/blob/main/test/configCases/process-assets/html-plugin/webpack.config.js

**Reply by alexander-akait** [reply]
Shorty -  you want apply information which you have after bundle, but DefinePlugin apply values before bundle

**Reply by melloware** [reply]
Ahhh so is it not possible?

**Reply by alexander-akait** [reply]
Using `DefinePlugin` it is impossible, but you can solve it using assets and replace content using `regexp` for example https://github.com/webpack/webpack/blob/main/test/configCases/process-assets/html-plugin/webpack.config.js#L107

**Comment by melloware** [comment, marked as answer]
I solved it by simply pregenerating my value up front and then passing to the Define plugin.  Thanks for your help!
